### PR TITLE
[Inductor] Refactor group/batch fusion to support user defined execution order and configs

### PR DESCRIPTION
**Pull request by yanboliang**
Meta internal customers need more flexible configs on these group/batch fusion's execution order and parameters, I'd like to provide a new inductor config that users can fine and auto tune these group/batch fusions for different models.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_inductor/fx_passes/pre_grad.py [hunk 1]
```diff
@@ -67,7 +67,7 @@ def pre_grad_passes(gm: torch.fx.GraphModule, example_inputs):
     if config.pattern_matcher:
         lazy_init()
         gm = fuse_fx(gm, example_inputs)
-        group_batch_fusion_pre_grad_passes(gm.graph)
+        group_batch_fusion_passes(gm.graph, True)
```
Comment (xuzhao9) [inline torch/_inductor/fx_passes/pre_grad.py:70]: It is clearer to specify `pre_grad=True` here.

File: torch/_inductor/fx_passes/post_grad.py [hunk 2]
```diff
@@ -76,7 +76,7 @@ def post_grad_passes(gm: torch.fx.GraphModule, is_inference: bool):
     if config.pattern_matcher:
         lazy_init()
 
-        group_batch_fusion_post_grad_passes(gm.graph)
+        group_batch_fusion_passes(gm.graph, False)
```
Comment (xuzhao9) [inline torch/_inductor/fx_passes/post_grad.py:79]: It is clearer to specify `pre_grad=False` here.

File: torch/_inductor/fx_passes/group_batch_fusion.py [hunk 3]
```diff
@@ -664,29 +710,29 @@ def print_graph(graph: torch.fx.Graph, msg: str):
...
+    print_graph(graph, "Before group_batch fusion in post grads pass.")
+    fusions: List[GroupBatchFusionBase] = []
 
+    if pre_grad:
+        fusions = generate_fusion_from_config(
+            config.pre_grad_fusion_options, pre_grad=True
+        )
+    elif has_fbgemm:  # Only group fusion (which needs fbgemm) in post grad.
+        fusions = generate_fusion_from_config(
+            config.pre_grad_fusion_options, pre_grad=False
```
Comment (xuzhao9) [inline torch/_inductor/fx_passes/group_batch_fusion.py:733]: Just wondering should this be `config.post_grad_fusion_options`?

File: torch/_inductor/config.py [hunk 4]
```diff
@@ -85,12 +85,26 @@
...
 group_fusion = False
 
-# enable pattern match with batch fusion (using torch op)
+# Deprecated
 batch_fusion = True
 
+# Pre grad group/batch fusion and options in order, set to empty dict to disable fusion.
+# Call `torch._inductor.fx_passes.group_batch_fusion.list_group_batch_fusions()` to see available fusions.
+pre_grad_fusion_options: Dict[str, Dict[str, Any]] = {
+    "batch_linear": {},
```
Comment (xuzhao9) [inline torch/_inductor/config.py:97]: Just want to confirm, the fusions listed in this `Dict` are enabled by default, is that correct?

File: torch/_inductor/fx_passes/group_batch_fusion.py (hunk 3 above)
Comment (yanboliang) [reply torch/_inductor/fx_passes/group_batch_fusion.py]: Yes, thanks for catching this!

File: torch/_inductor/config.py (hunk 4 above)
Comment (yanboliang) [reply torch/_inductor/config.py:97]: You are right!

**Comment by xuzhao9** [comment]
Looks like there is a conflict with the main branch, any progress on this?

**Review by xuzhao9 — APPROVED** [review]
LGTM

**Pull request closed without merging**